### ISSUE TITLE
Monal loses connection while in background, slow message sync, no push notifications

### ISSUE THREAD

**Issue by bytehamm3r** [monal-im/Monal#676]
### Description of my issue or feature request:
Hello,

I had my setup running for the past months without any problem with Monal (iOS). My server configuration worked out very well.

Now, since the roll out of Monal version 5.0 I experienced that all contacts using Monal are not persistent online anymore. I didn't find any error in the server logs so far. Instead I've identified that Monal loses its connection when running in the background which it did not before version 5.0. Even more there are no push notifications anymore although Monal gets registered to the push server in a correct manner.

I tried to check if any changes in Monal 5.0 requires a reconfiguration of the server but I haven't found anything related. This thesis would also contradict to the situation that Gajim and Conversations are running flawlessly.

In conclusion something had gone wrong with the new version - does anyone have some advice? A lot of my contacts are now practically offline.

### System information (if you can provide):

iOS version: 5.0 (750)
macOS version: -
Used XMPP server (domain): not public
Used XMPP server software & version: prosody 0.11.9
How many accounts are registered in your Monal app: -

---
[X] I have cross-check this overview https://github.com/anurodhp/Monal/issues/322 as well es filtered for related labels https://github.com/anurodhp/Monal/labels

**Comment by tmolitor-stud-tu** [comment]
Monal *never* had an idle TCP connection open when it was in the background. Apple does not allow this for apps and freezes apps after being active in the background for a maximum of 30 seconds.

Does your server provide a XEP-198 stream management session and what timeout is configured for these stream management sessions? (mod_smacks in prosody).

Can you provide a screenshot of settings --> notifications? that would help a lot here.

**Comment by bytehamm3r** [comment]
In the view of the network this may be correct, but in the result the contacts are shown as offline in other clients and the communication can no longer take place. This is meant with "loses connection".

Of course 'mod_smacks' is part of my server. The configuration of it is completely default.

The iOS notification configuration is sufficiently on multiple devices. Notifications allowed, Styles activated and so on.

Once again, the configuration of the server as well all the iOS clients haven't changed at all. Everything worked until the update to version 5.0.

**Comment by Echolon** [comment]
@bytehamm3r at its best really provide the screenshots. Its the best way to exclude potential configuration issues.

If you can, you can also reproduce with active logs and send those to the developers.

**Comment by tmolitor-stud-tu** [comment]
so your mod_smacks timeout is 5 minutes and clients are shown as "offline" in gajim etc. after this.
That's perfectly okay and absolutely normal.

once you write a message to those accounts a push will be sent out to monal's appserver and that will notify apples's servers to send out a silent push to the device which will wake up the client and let it connect to your xmpp server retrieving the message through the mam (XEP-0313 archive management).

That means the only symptom you are experiencing is push not working. Hence my question to send a screenshot of Settings --> Notifications to help diagnose that problem.

You can send me the "node" and "secret" values used when registering/activating push on your server and I can try to manually send out a push to the device to check if that works okay.
You can send those two values out of band to thilo+monal@eightysoft.de

**Comment by bytehamm3r** [comment]
This is what it looks like when I open the app while it was running in the background for some time (red dot). After a few seconds, the dot goes green and I appear visible in other clients. Besides this, I erased the account information for privacy reasons.

![pic1](https://user-images.githubusercontent.com/77917421/124016279-c1608b00-d9e5-11eb-9cbe-31d09d5adb63.jpeg)


In contrast when online:

![pic](https://user-images.githubusercontent.com/77917421/124170166-6000f080-daa7-11eb-97d2-ea7f2b760b20.jpg)

**Comment by bytehamm3r** [comment]
@tmolitor-stud-tu are these values within the server log files?

**Comment by tmolitor-stud-tu** [comment]
I meant the menu point "Notifications" in the Monal settings, not the iOS settings. You can see the menu entry in your first screenshot.

yes, these values are visiblein the server logsif you have debug logging on. But you can extract them from the file storage or mysql storage backend as well.
You can extract them from /var/lib/prosody/<your server dns>/cloud_notify/<username>.dat

**Comment by bytehamm3r** [comment]
Ok, now the correct screenshot…

![pic3](https://user-images.githubusercontent.com/77917421/124016490-fcfb5500-d9e5-11eb-936b-2b5310e02e16.jpeg)

**Comment by tmolitor-stud-tu** [comment]
Please go to Settings --> Logs and tap "Reconnect all".
Then wait until your red dot turns green and take a new screenshot of Settings --> Notifications

**Comment by bytehamm3r** [comment]
Here is the prosody log file while using Monal...
```
...
Jun 30 13:49:41 c2s555fb****** info    Client connected
Jun 30 13:49:43 c2s555fb****** info    Stream encrypted (TLSv1.2 with ECDHE-RSA-AES256-GCM-SHA384)
Jun 30 13:49:43 c2s555fb****** info    Authenticated as user@domain.tld
Jun 30 13:49:47 s2sout555fb3******      info    Stream encrypted (TLSv1.3 with TLS_AES_256_GCM_SHA384)
Jun 30 13:49:47 s2sout555fb3******      info    Outgoing s2s connection domain.tld->ios13push.monal.im complete
Jun 30 13:50:07 c2s555fb3****** info    Client disconnected: closed
*** Monal now open ***
Jun 30 13:51:40 c2s555fb3****** info    Client connected
Jun 30 13:51:41 c2s555fb3****** info    Stream encrypted (TLSv1.2 with ECDHE-RSA-AES256-GCM-SHA384)
Jun 30 13:51:42 c2s555fb3****** info    Authenticated as user@domain.tld
*** Monal to background ***
Jun 30 13:52:16 c2s555fb3****** info    Client disconnected: closed
*** Monal now open ***
Jun 30 13:59:15 c2s555fb3****** info    Client connected
Jun 30 13:59:16 c2s555fb3****** info    Stream encrypted (TLSv1.2 with ECDHE-RSA-AES256-GCM-SHA384)
Jun 30 13:59:16 c2s555fb3****** info    Authenticated as user@domain.tld
*** Monal to background ***
Jun 30 14:00:06 c2s555fb3****** info    Client disconnected: closed
...
```

**Comment by bytehamm3r** [comment]
> Please go to Settings --> Logs and tap "Reconnect all".
> Then wait until your red dot turns green and take a new screenshot of Settings --> Notifications

I don't find that... I've no "Logs" entry in the settings.

**Comment by tmolitor-stud-tu** [comment]
> I don't find that... I've no "Logs" entry in the settings.

It's the second entry in the settings menu if counted from bottom upwards.

**Comment by tmolitor-stud-tu** [comment]
The prosody log looks perfectly reasonable and is completely normal.

Would be nice to be able to manually send a push and see if this works.

**Comment by tmolitor-stud-tu** [comment]
forget it, the logs menu entry is only present in beta and nightly builds...sorry. then just wait until the dot is green before taking the screenshot.

**Comment by bytehamm3r** [comment]
@tmolitor-stud-tu Ok. I need to extract the requested values from mysql. I'll get in touch with you by the weekend at lastest. I already wrote down the contact informations. Thank you so far.

**Comment by Echolon** [comment]
@bytehamm3r please dont paste sensitive data here. you can forward it directly to Thilos mail

thilo+monal (at) eightysoft (dot) de

**Comment by roman901** [comment]
![IMG_E59F247962D4-1](https://user-images.githubusercontent.com/2963008/124139235-178a0880-da90-11eb-8830-23215be979db.jpeg)
Can agree - no pushes, fast disconnect after screen locking. 
But, on iPad I have Monal 4.9.2 and can reproduce this. Open app - tap to settings - connected right now
Switched to springboard and re-open app - red dot -> green dot - connected right now. And also no pushes.
iOS 14.6

**Comment by tmolitor-stud-tu** [comment]
@roman901 what is "fast disconnect"? after screen locking the device should still be connected for about 20 seconds...if this isn't the case: could you install the beta build and verify if this is the case there, too?

And if the beta build behaves the same: can you send me a logfile via email (please don't post it here): thilo+monal@eightysoft.de

go to settings -> logs -> tap the leftmost icon on the bottom bar and send me the resulting file
please add the exact time you put the app into background to that email, too, so that I'm able to find the corresponding part in the logs.

**Comment by tmolitor-stud-tu** [comment]
Some more questions for both of you:

1. Do you have battery saving mode turned on on your iphone?
2. do you have mod_csi_battery_saver or mod_csi_simple activated on your prosody server (or some similar module on your ejabberd)?
3. do you have SRV records for your server?
4. how mny contacts do you have in your roster?
5. how many mucs do you have bookmarked/joined (just count them in your contact list), channels and groups
6. do you get "could not synchronize" notifications from monal, if so, how often?
7. how long is the tcp connection from monal to your server open, after putting the app into the background? (should be ~20 seconds to be correct)

**Comment by tmolitor-stud-tu** [comment]
It seems sometimes the apple push mechanism deadlocks and after a device restart pushes come in again.
Could you restart your phone to check if that is your problem?

**Comment by roman901** [comment]
After updating to 5.0 beta all problems gone, pushes are coming and connection is stable

**Comment by tmolitor-stud-tu** [comment]
@roman901 interesting: the current beta an stable builds are completely identical (with about 10 minutes between the builds and no code change).

I've added the Logs menu entry back in for the next stable release (Monal 5.0.1) to help debugging issues on stable (I never wanted to remove it in the first place, my mistake).
Maybe we get some clues what's going on in the stable by reading the logs once 5.0.1 got released.

**Comment by roman901** [comment]
@tmolitor-stud-tu yes, I already thinking about that. If it happens again, I will try to get logs. Also, no devices restarts happened, only application reinstalls

**Comment by tmolitor-stud-tu** [comment]
@roman901 can you check if your stable gets resposive again after restarting your device?

**Comment by tmolitor-stud-tu** [comment]
@Echolon https://github.com/monal-im/Monal/commit/039d9f77f7e9c4b717d4cc2c1ed59c7ed798435f

**Comment by bytehamm3r** [comment]
@tmolitor-stud-tu Related to your questions:

1. No
2. No. Couldn't get any CSI module running without breaking push notifications, so none is running since I configured the server initially.
3. Yes.
4. It's not related to the quantity of contacts. All users with Monal are effected equally on my server.
5. None, deactivated on my server.
6. Sporadically/phasically a few times a day. Currently none.
7. Seems to be around 20s.

Restarting the device doesn't change anything.

At this point, I would like to clarify that the missing push notifications are merely a symptom. The root cause of all is the disconnection of the Monal client which can be seen in a offline status in the contact list. We should focus on the connection loss shortly after locking the iPhone.

**Comment by tmolitor-stud-tu** [comment]
@bytehamm3r having no csi module is a problem: without csi your server will generate many pushes and apple will most likely throttle them at some point...that means pushes will be filtered out by apple for some days before they eventually will come through again (could be a long time though).

I asked question 4 because having more active contacts without csi will create more pushes, thus earlier throttling by apple. But without csi only a few contacts could already be enough for apple to throttle pushes.


**_I want to stress this here again, because there seems to be some misunderstanding:_**
Apple does **NOT** allow apps to stay connected when not in foreground.
That means Monal disconnecting after 20 seconds is perfectly right. After the TCP connection is closed your server will keep the session alive for another 600 seconds (that's done by mo_smacks) and after that the client will **APPEAR** offline.
The old concept of "online" meaning "connected" is somewhat deprecated when talking about the mobile world.

But that does not mean a user won't get any messages once the smacks session timed out: mod_cloud_notify will send pushes for new messages written to the MAM archive and Monal will happily connect upon those pushes and fetch the missing stuff from MAM.
_If you have enabled MAM on your server of course._
If you don't have MAM enabled, there is nothing I can do to make Monal receive new messages, short of coming over to you and opening Monal on your phone myself ;)

**Some more thoughts:**
Normally mod_cloud_notify and mod_smacks will interact in a way that makes the smacks timeout last longer if no push was sent out yet.That essentially makes the 600 second timeout of mod_smacks begin when the first push gets sent instead of when the TCP connection was closed.
If you'd have a CSI module enabled, that would essentially prevent such frequent pushes and make the smacks session last longer without apple throtling the pushes, potentially several days).
Longer lasting smacks session will let your clients **APPEAR* online for longer times, but as I said earlier: this notion of online/offline is not meaningful for mobile clients at all.

_I'd recommend to use mod_csi_battery_saver working good with Conversations, Monal, Dino and any other client I know of._

**Comment by bytehamm3r** [comment]
@tmolitor-stud-tu

I'm using MAM and Smacks as supposed to be.

As mentioned: I have no good experience with CSI with push notifications. Evenmore, I had no throttling yet.

**Comment by tmolitor-stud-tu** [comment]
@bytehamm3r From your logs and sql database I can see you don't have mod_cloud_notify activated. Without that module you won't ever get a push. I thought you'd have that activated and never asked you about that, but apparently you have not.

You will need mod_csi_battery_saver if you don't want your battery to drain very fast and/or apple throttle pushes to your device.

**Comment by bytehamm3r** [comment]
@tmolitor-stud-tu Well, I can see it is activated...
```

29898:Jul 03 11:05:03 pubsub.domain.tld:tls	info	Certificates loaded
29899:Jul 03 11:05:03 pubsub.domain.tld:tls	info	Certificates loaded
29980:Jul 03 11:05:04 domain.tld:cloud_notify	info	Module loaded
29981:Jul 03 11:05:04 domain.tld:cloud_notify	info	Module loaded
30197:Jul 03 11:05:08 c2s55f830******	debug	Received[c2s]: <iq id='...' type='get' to='pubsub.domain.tld'>
30198:Jul 03 11:05:08 c2s55f830******	debug	Sending[c2s]: <iq id='... type='result' to='me@domain.tld/gajim' from='pubsub.domain.tld'>
30750:Jul 03 11:05:11 c2s55f830******	debug	Received[c2s]: <iq id='...' type='get' to='pubsub.domain.tld'>
30751:Jul 03 11:05:11 c2s55f830******	debug	Sending[c2s]: <iq id='...' type='result' to='me@domain.tld/notebook' from='pubsub.domain.tld'>
31123:Jul 03 11:07:52 c2s55f830******	debug	Received[c2s]: <iq id='...' type='set' to='ios13push.monal.im'>
31126:Jul 03 11:07:52 s2sout55f830******	debug	First attempt to connect to ios13push.monal.im, starting with SRV lookup...
31127:Jul 03 11:07:52 adns	debug	Records for _xmpp-server._tcp.ios13push.monal.im. not in cache, sending query (thread: 0x55f830d0d160)...
31150:Jul 03 11:07:52 c2s55f830******	debug	Received[c2s]: <iq id='...' type='get' to='pubsub.domain.tld'>
31151:Jul 03 11:07:52 c2s55f830******	debug	Sending[c2s]: <iq id='...' type='result' to='me@domain.tld/Monal-iOS' from= pubsub.domain.tld'>
31244:Jul 03 11:07:52 adns	debug	Reply for _xmpp-server._tcp.ios13push.monal.im. (thread: 0x55f830d0d160)
31245:Jul 03 11:07:52 s2sout55f830******	debug	ios13push.monal.im has no SRV records, falling back to A/AAAA
31246:Jul 03 11:07:52 adns	debug	Records for ios13push.monal.im not in cache, sending query (thread: 0x55f830cf2920)...
31248:Jul 03 11:07:52 adns	debug	Records for ios13push.monal.im not in cache, sending query (thread: 0x55f830c25660)...
31253:Jul 03 11:07:52 adns	debug	Reply for ios13push.monal.im (thread: 0x55f830cf2920)
31254:Jul 03 11:07:52 s2sout55f830******	debug	DNS reply for ios13push.monal.im gives us 34.227.106.56
31256:Jul 03 11:07:52 adns	debug	Reply for ios13push.monal.im (thread: 0x55f830c25660)
31257:Jul 03 11:07:52 s2sout55f830******	debug	Beginning new connection attempt to ios13push.monal.im ([34.227.106.56]:5269)
31260:Jul 03 11:07:52 s2sout55f830******	debug	Sending[s2sout_unauthed]: <stream:stream xmlns:db='jabber:server:dialback' xmlns:stream='http://etherx.jabber.org/streams' xml:lang='en' from='domain.tld' to='ios13push.monal.im' version='1.0' xmlns='jabber:server'>
31264:Jul 03 11:07:52 domain.tld:tls	debug	ios13push.monal.im is offering TLS, taking up the offer...
31272:Jul 03 11:07:52 s2sout55f830******	debug	Sending[s2sout_unauthed]: <stream:stream xmlns:db='jabber:server:dialback' xmlns:stream='http://etherx.jabber.org/streams' xml:lang='en' from='domain.tld' to='ios13push.monal.im' version='1.0' xmlns='jabber:server'>
31278:Jul 03 11:07:52 x509	debug	Cert dNSName ios13push.monal.im matched hostname
31282:Jul 03 11:07:52 domain.tld:saslauth	debug	Initiating SASL EXTERNAL with ios13push.monal.im
31285:Jul 03 11:07:52 domain.tld:saslauth	debug	SASL EXTERNAL with ios13push.monal.im succeeded
31287:Jul 03 11:07:52 s2sout55f830******	debug	Sending[s2sout_unauthed]: <stream:stream xmlns:db='jabber:server:dialback' xmlns:stream='http://etherx.jabber.org/streams' xml:lang='en' from='domain.tld' to='ios13push.monal.im' version='1.0' xmlns='jabber:server'>
31288:Jul 03 11:07:52 s2sout55f830******	debug	connection domain.tld->ios13push.monal.im is now authenticated for ios13push.monal.im
31291:Jul 03 11:07:52 s2sout55f830******	info	Outgoing s2s connection domain.tld->ios13push.monal.im complete
31292:Jul 03 11:07:52 s2sout55f830******	info	Outgoing s2s connection domain.tld->ios13push.monal.im complete
```

**Comment by tmolitor-stud-tu** [comment]
@bytehamm3r I think I've figured out what your problem is: your server does not allow incoming s2s connections from the monal push server. But you need that to transfer the push node and secret values to Monal so that it can use them to register for push.

Monal 4.9.2 used an out of band channel for that (http), Monal 5.0 now uses xmpp for that as well (in line with what conversations is doing).

**Comment by tmolitor-stud-tu** [comment]
Regarding the csi module: I'm the author of mod_csi_battery_saver. If you tell me whats wrong with that module, I can fix it. That would save a huge amount of your battery on your device.

**Comment by tmolitor-stud-tu** [comment]
@roman901 maybe csi is not activated on your server as well, could be apple is throttling you for the stable app, but not for the beta (yet).

**Comment by tmolitor-stud-tu** [comment]
I'm closing this as solved, @roman901 feel free to reopen it if your problems resurrect again and can't be fixed with the above.
